File: codigo/15_pandeo/calculo_pandeo_eigenproblema.m
clear all; close all; clc;

% Raíces precalculadas de tan(x)=x para el caso empotrado-articulado
phi = [ 4.4934094579, 7.7252518369, 10.9041216594, 14.0661939128 ];

% Raíces de 2 - kL*sin(kL) - 2cos(kL) = 0 para el caso empotrado-empotrado
knL_ee = [ 2*pi, 2*phi(1), 4*pi, 2*phi(2) ];

%% Se definen los parámetros del problema
N = 100;                % Número de nodos internos en la discretización
E = 200e9;              % [Pa]  Módulo de Young para el acero (200 GPa)
I = (0.01)^4/12;        % [m^4] Momento de inercia (sección cuadrada 1cm x 1cm)
L = 1.0;                % [m]   Longitud de la viga

% Casos de condiciones de apoyo
casos = {'articulado-articulado', 'empotrado-empotrado', ...
         'empotrado-libre', 'empotrado-articulado'};

%% Resolver y graficar para cada caso de condiciones de apoyo
for c = 1:length(casos)
    caso = casos{c};
    fprintf('\n--- CASO %s ---\n', upper(caso));
    
    % Obtener solución numérica
    [P_cr_numerica, modos_pandeo] = resolver_pandeo(caso, N, L, E, I);
    
    % Calcular soluciones analíticas
    P_cr_analitica = zeros(4, 1);
    for n = 1:4
        P_cr_analitica(n) = P_cr_analitico(caso, n, L, E, I, phi, knL_ee);
    end
    
    % Tabla comparativa
    fprintf('%4s | %20s | %20s | %10s\n', 'Modo', 'P_cr Numérica (kN)', ...
            'P_cr Analítica (kN)', 'Error (%)');
    fprintf('%s\n', repmat('-', 1, 65));
    
    error = 100*abs(P_cr_numerica - P_cr_analitica)./P_cr_analitica;
    for n = 1:4
        fprintf('%4d | %20.5f | %20.5f | %10.5f\n', n, ...
                P_cr_numerica(n)/1000, P_cr_analitica(n)/1000, error(n));
    end

    % Se grafica la comparación analítica vs numérica de los modos de pandeo
    figure;
    sgtitle(sprintf('Primeros cuatro modos de pandeo: %s', caso));    
    for n = 1:4
        subplot(2, 2, n);
        hold on; grid on;
                      
        % Solución numérica (línea azul con puntos)
        x_num = linspace(0, L, N+2);
        if strcmp(caso, 'empotrado-libre')
            y_num = [0; modos_pandeo(:, n)];
        else
            y_num = [0; modos_pandeo(:, n); 0];
        end
        plot(x_num, y_num, 'bo-', 'MarkerSize', 3, 'DisplayName', 'Numérica (MDF)');
        
        % Solución analítica (línea roja punteada)
        x_an = linspace(0, L, 200);
        y_an = modo_pandeo_analitico(caso, x_an, n, L, phi, knL_ee);
        
        % Normalizar y graficar la forma analítica
        y_an_normalizada = y_an/max(abs(y_an));

        % Voltear si es necesario para coincidir con la orientación del modo numérico
        % Si y_an_normalizada y y_num tienen la misma orientación >> productos positivos
        % Si tienen orientaciones opuestas >>> productos negativos
        if sum(y_an_normalizada .* interp1(x_num, y_num, x_an)) < 0
            y_an_normalizada = -y_an_normalizada;
        end
        
        plot(x_an, y_an_normalizada, 'r--', 'DisplayName', 'Analítica');
        
        title(sprintf('Modo %d', n));
        grid on;        
        xlabel('Posición a lo largo de la columna (m)');
        ylabel('Deflexión Normalizada');
        legend('Location', 'best');
    end
end

%% Cálculo de carga crítica analítica
function P_cr = P_cr_analitico(cond_apoyo, n, L, E, I, phi, knL_ee)
    switch cond_apoyo
        case 'articulado-articulado'
            P_cr = (n^2 * pi^2 * E*I)/L^2;
        case 'empotrado-empotrado'
            P_cr = (knL_ee(n)^2 * E*I)/L^2;
        case 'empotrado-libre'
            P_cr = ((2*n - 1)^2 * pi^2 * E*I)/(4*L^2);
        case 'empotrado-articulado'
            P_cr = (phi(n)^2 * E*I)/L^2;
        otherwise
            error('Tipo de condición de apoyo desconocida');
    end
end

%% Modo de pandeo analítico
function vx = modo_pandeo_analitico(cond_apoyo, x, n, L, phi, knL_ee)
    switch cond_apoyo
        case 'articulado-articulado'
            vx = sin(n*pi*x/L);
        case 'empotrado-empotrado'
            knL = knL_ee(n);
            kn = knL/L;
            if mod(n, 2) ~= 0  % n = 1, 3, 5, 7, ...
                % Forma modal simplificada
                vx = cos(kn*x)- 1;
            else               % n = 2, 4, 6, 8, ...
                % Forma modal general
                vx = cos(kn*x) - 1 ... 
                     - ((sin(kn*x) - kn*x)/(sin(knL) - knL))*(cos(knL) - 1);
            end
        case 'empotrado-libre'
            vx = cos((2*n - 1)*pi*x/(2*L)) - 1;
        case 'empotrado-articulado'
            kn = phi(n)/L;
            vx = sin(kn*x) - kn*L*cos(kn*x) + kn*(L - x);
        otherwise
            error('Tipo de condición de apoyo desconocida');
    end
end

%% Resolver el problema de pandeo
function [P_cr, modos_pandeo] = resolver_pandeo(cond_apoyo, N, L, E, I)
    % Resuelve el problema de pandeo de columna para los primeros cuatro modos
    % usando el método de las diferencias finitas.

    % Calcula numéricamente las cargas críticas y los modos de pandeo asociados

    % Definición del paso de la malla
    h = L/(N + 1);
    
    % Extraer condiciones de apoyo izquierda y derecha
    apoyo = split(cond_apoyo, '-');
    apoyo_izq = apoyo{1};
    apoyo_der = apoyo{2};
    
    % Construcción de matrices de rigidez Ke y Kg
    if strcmp(apoyo_der, 'libre')
        Ke = zeros(N+1);
        Kg = zeros(N+1);
    else
        Ke = zeros(N);
        Kg = zeros(N);
    end
    
    % Condición de apoyo izquierda
    if strcmp(apoyo_izq, 'articulado')
        Ke(1, 1:3) = [5, -4, 1];
        Ke(2, 1:4) = [-4, 6, -4, 1];
        Kg(1, 1:2) = [2, -1];
    elseif strcmp(apoyo_izq, 'empotrado')
        Ke(1, 1:3) = [7, -4, 1];
        Ke(2, 1:4) = [-4, 6, -4, 1];
        Kg(1, 1:2) = [2, -1];
    else
        error('Condición de apoyo izquierda desconocida');
    end
    
    % Discretización de la ecuación diferencial (diferencias finitas centradas)
    if strcmp(apoyo_der, 'libre')
        for i = 3:N-1
            Ke(i, i-2:i+2) = [1, -4, 6, -4, 1];
        end
        for i = 2:N
            Kg(i, i-1:i+1) = [-1, 2, -1];
        end
    else
        for i = 3:N-2
            Ke(i, i-2:i+2) = [1, -4, 6, -4, 1];
        end
        for i = 2:N-1
            Kg(i, i-1:i+1) = [-1, 2, -1];
        end
    end
    
    % Condición de apoyo derecha
    switch apoyo_der
        case 'articulado'
            Ke(N,   N-2:N  ) = [1, -4, 5];
            Ke(N-1, N-3:N  ) = [1, -4, 6, -4];
            Kg(N,   N-1:N  ) = [-1, 2];
        case 'empotrado'
            Ke(N,   N-2:N  ) = [1, -4, 7];
            Ke(N-1, N-3:N  ) = [1, -4, 6, -4];
            Kg(N,   N-1:N  ) = [-1, 2];
        case 'libre'
            Ke(N+1, N-1:N+1) = [1, -2, 1];
            Ke(N,   N-2:N+1) = [1, -4, 5, -2];
            Kg(N+1, N  :N+1) = [-1, 1];
        otherwise
            error('Condición de apoyo derecha desconocida');
    end
    
    % Escalar las matrices de rigidez
    Ke = Ke * (E*I/h^4);
    Kg = Kg * (1/h^2);
    
    % Resolver problema de eigenvalores generalizado
    [vecp, valp] = eig(Ke, Kg);
     
    % Ordenar valores propios
    [valp, idx] = sort(diag(valp), 'ascend');
    vecp = vecp(:, idx);
    
    % Seleccionar los primeros 4 modos positivos
    idx_positivos = valp > 1e-6;
    valp_pos = valp(idx_positivos);
    vecp_pos = vecp(:, idx_positivos);
    num_modos = min(4, length(valp_pos));

    P_cr = valp_pos(1:num_modos);
    modos_pandeo = vecp_pos(:, 1:num_modos);
    
    % Normalizar modos de pandeo y asegurar orientación consistente
    for i = 1:size(modos_pandeo, 2)
        modos_pandeo(:, i) = modos_pandeo(:, i)/max(abs(modos_pandeo(:, i)));
        if sum(modos_pandeo(:, i)) < 0
            modos_pandeo(:, i) = -modos_pandeo(:, i);
        end
    end
end

File: codigo/15_pandeo/calculo_pandeo_eigenproblema.py
import numpy as np
from scipy.linalg import eigh
import matplotlib.pyplot as plt

# Raíces precalculadas de tan(x)=x para el caso empotrado-articulado
phi = [np.nan, 4.4934094579, 7.7252518369, 10.9041216594, 14.0661939128]

# Raíces de 2 - kL*sin(kL) - 2cos(kL) = 0 para el caso empotrado-empotrado
knL_ee = [np.nan, 2*np.pi, 2*phi[1], 4*np.pi, 2*phi[2]]

# %% Cálculo de carga crítica analítica
def P_cr_analitico(cond_apoyo, n, L=1.0, E=1.0, I=1.0):
    match cond_apoyo:
        case 'articulado-articulado':
            P_cr = (n**2 * np.pi**2 * E*I)/L**2
        case 'empotrado-empotrado':
            P_cr = (knL_ee[n]**2 * E*I)/L**2
        case 'empotrado-libre':
            P_cr = ((2*n-1)**2 * np.pi**2 * E*I)/(4*L**2)
        case 'empotrado-articulado':
            P_cr = (phi[n]**2 * E*I)/L**2
        case _:
            raise ValueError("Tipo de condición de apoyo desconocida")
    return P_cr

# %% Modo de pandeo analítico
def modo_pandeo_analitico(cond_apoyo, x, n, L=1.0):
    match cond_apoyo:
        case 'articulado-articulado':
            vx = np.sin(n*np.pi*x/L)
        case 'empotrado-empotrado':
            knL = knL_ee[n]
            kn = knL/L
            if n%2 != 0:  # n = 1, 3, 5, 7, ...
                # Forma modal simplificada
                vx = np.cos(kn*x) - 1
            else:         # n = 2, 4, 6, 8, ...
                # Forma modal general
                vx = np.cos(kn*x) - 1 - \
                    ((np.sin(kn*x) - kn*x)/(np.sin(knL) - knL))*(np.cos(knL) - 1)
        case 'empotrado-libre':
            vx = np.cos((2*n - 1)*np.pi*x/(2*L)) - 1
        case 'empotrado-articulado':
            kn = phi[n]/L
            vx = np.sin(kn*x) - kn*L*np.cos(kn*x) + kn*(L - x)
        case _:
            raise ValueError("Tipo de condición de apoyo desconocida")
    return vx

# %% Resolver el problema de pandeo
def resolver_pandeo(cond_apoyo, N, L=1.0, E=1.0, I=1.0):
    """
    Resuelve el problema de pandeo de columna para los primeros cuatro modos
    usando el método de las diferencias finitas.

    Calcula numéricamente las cargas críticas y los modos de pandeo asociados
    """
    # Definición del paso de la malla
    h = L/(N + 1)

    # Extraer condiciones de apoyo izquierda y derecha
    apoyo_izq, apoyo_der = cond_apoyo.split('-')

    # Construcción de las matrices de rigidez elástica y geométrica Ke y Kg
    if apoyo_der == 'libre':
        Ke = np.zeros((N+1, N+1))
        Kg = np.zeros((N+1, N+1))
    else:
        Ke = np.zeros((N, N))
        Kg = np.zeros((N, N))

    # Condición de apoyo izquierda
    match apoyo_izq:
        case 'articulado':
            Ke[0, [0, 1, 2   ]] = [5, -4, 1]
            Ke[1, [0, 1, 2, 3]] = [-4, 6, -4, 1]
            Kg[0, [0, 1      ]] = [2, -1]
        case 'empotrado':
            Ke[0, [0, 1, 2   ]] = [7, -4, 1]
            Ke[1, [0, 1, 2, 3]] = [-4, 6, -4, 1]
            Kg[0, [0, 1      ]] = [2, -1]
        case _:
            raise ValueError("Condición de apoyo izquierda desconocida")

    # Discretización de la ecuación diferencial con diferencias finitas
    # para el nodo interior "i"
    if apoyo_der == 'libre':
        for i in range(2, N-1):
            Ke[i, i-2:i+3] = [1, -4, 6, -4, 1]
        for i in range(1, N):
            Kg[i, i-1:i+2] = [-1, 2, -1]
    else:
        for i in range(2, N-2):
            Ke[i, i-2:i+3] = [1, -4, 6, -4, 1]
        for i in range(1, N-1):
            Kg[i, i-1:i+2] = [-1, 2, -1]

    # Condición de apoyo derecha
    match apoyo_der:
        case 'articulado':
            Ke[N-1, [     N-3, N-2, N-1]] = [1, -4, 5]
            Ke[N-2, [N-4, N-3, N-2, N-1]] = [1, -4, 6, -4]
            Kg[N-1, [          N-2, N-1]] = [-1, 2]
        case 'empotrado':
            Ke[N-1, [     N-3, N-2, N-1]] = [1, -4, 7]
            Ke[N-2, [N-4, N-3, N-2, N-1]] = [1, -4, 6, -4]
            Kg[N-1, [          N-2, N-1]] = [-1, 2]
        case 'libre':
            Ke[  N, [     N-2, N-1, N  ]] = [1, -2, 1]
            Ke[N-1, [N-3, N-2, N-1, N  ]] = [1, -4, 5, -2]
            Kg[  N, [          N-1, N  ]] = [-1, 1]
        case _:
            raise ValueError("Condición de apoyo derecha desconocida")

    # Escalar las matrices de rigidez
    Ke *= E*I/h**4
    Kg *= 1/h**2

    # Resolver problema de eigenvalores generalizado
    # Con eigh() los valores propios ya son reales y están ordenados
    valp, vecp = eigh(Ke, Kg)

    # Seleccionar los primeros 4 modos de pandeo
    P_cr         = valp[  :4]
    modos_pandeo = vecp[:,:4]

    # Normalizar los modos de pandeo
    for i in range(4):
        modos_pandeo[:, i] /= np.max(np.abs(modos_pandeo[:, i]))

    return P_cr, modos_pandeo

# Se definen los parámetros del problema
N = 50          # Número de nodos internos en la discretización
E = 200e9       # [Pa]  Módulo de Young para el acero (200 GPa)
I = 0.01**4/12  # [m^4] Momento de inercia (sección cuadrada 1cm x 1cm)
L = 1.0         # [m]   Longitud de la viga

# Casos de condiciones de apoyo
casos = [ 'articulado-articulado', 'empotrado-empotrado', 
          'empotrado-libre', 'empotrado-articulado'      ]

# %% Resolver y graficar para cada caso de condiciones de apoyo
for caso in casos:
    print(f"\n--- CASO {caso.upper()} ---")

    # Obtener solución numérica
    P_cr_numerica, modos_pandeo = resolver_pandeo(caso, N=N, L=L, E=E, I=I)

    # Calcular soluciones analíticas
    P_cr_analitica = np.zeros(4)
    for i in range(4):
        P_cr_analitica[i] = P_cr_analitico(caso, n=i+1, L=L, E=E, I=I)

    # Tabla comparativa entre Pcr numérica y analítica
    print(f"{'Modo':>4} | {'P_cr Numérica (kN)':>20} | "
          f"{'P_cr Analítica (kN)':>20} | {'Error (%)':>10}")
    print("-" * 65)
    error = 100*abs(P_cr_numerica - P_cr_analitica)/P_cr_analitica
    for i in range(4):
        print(f"{i+1:>4} | {P_cr_numerica[i]/1000:>20.5f} | "
              f"{P_cr_analitica[i]/1000:>20.5f} | {error[i]:>10.5f}")

    # Se grafica la comparación analítica vs numérica de los modos de pandeo
    fig, axs = plt.subplots(2, 2, figsize=(12, 8), sharex=True)
    fig.suptitle(f'Primeros cuatro modos de pandeo: {caso.title()}', fontsize=16)

    for i in range(4):
        ax = axs.flat[i]

        # Solución analítica (línea azul)
        x = np.linspace(0, L, N+2)
        y_an = modo_pandeo_analitico(caso, x, n=i+1, L=L)

        # Normalizar y graficar la forma analítica
        y_an_normalizada = y_an/np.max(np.abs(y_an))
        ax.plot(x, y_an_normalizada, 'b-', label='Solución analítica')

        # Solución numérica (puntos rojos)
        if caso == 'empotrado-libre':
            y_num = np.hstack([0, modos_pandeo[:, i]])
        else:
            y_num = np.hstack([0, modos_pandeo[:, i], 0])

        # Voltear si es necesario para coincidir con la orientación del modo
        # numérico. Si y_an y y_num tienen la misma orientación, los
        # productos son positivos; si tienen orientaciones opuestas, los
        # productos son negativos
        if np.sum(y_an * y_num) < 0:
            y_num *= -1
        ax.plot(x, y_num, 'r.', markersize=4, label='Diferencias finitas')

        ax.set_title(f'Modo de pandeo {i+1}')
        ax.grid(True)
        ax.set_ylabel("Deflexión Normalizada")

    axs.flat[0].legend()
    axs.flat[2].set_xlabel("Posición a lo largo de la columna (m)")
    axs.flat[3].set_xlabel("Posición a lo largo de la columna (m)")
    plt.tight_layout()
    plt.show()